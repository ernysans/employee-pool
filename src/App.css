.mdc-tab .mdc-tab__icon,
.mdc-tab .mdc-tab__text-label{
  color: #ffff;
}
.mdc-top-app-bar * {
  color: #ffff;
  --mdc-theme-primary: #ffff;
}
