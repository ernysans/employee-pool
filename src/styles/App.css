.mdc-tab .mdc-tab__icon,
.mdc-tab .mdc-tab__text-label {
  color: #ffff;
}

.mdc-tab-bar * {
  color: #ffff;
  --mdc-theme-primary: #ffff;
}

.mdc-list-item__graphic {
  width: 2rem;
  height: 2rem;
  border-radius: 50%;
  margin-right: 1rem;
}
.mdc-select__selected-text-container .mdc-list-item__graphic {
  width: 1.5rem;
  height: 1.5rem;
  margin-right: 1.5rem;
}
.mdc-select__menu .mdc-list-item {
  padding: 0.5rem 1rem;
}

.avatar {
  width: 2rem;
  height: 2rem;
  border-radius: 50%;
  margin-right: 1rem;
}

/* Shared styles */
.container {
  display: block;
  width: 100%;
  max-width: 1024px;
  margin: 0 auto;
  position: relative;
}

.content-container {
  padding: 1rem 0;
  margin: 0 1rem;
  display: block;
}
