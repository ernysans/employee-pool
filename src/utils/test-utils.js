import {BrowserRouter} from "react-router-dom";
import {render} from "@testing-library/react";
import {userEvent} from "@testing-library/user-event";
import {store} from "./store";
import {Provider} from "react-redux";

export const renderWithRouter = (ui, {route = '/'} = {}) => {
  window.history.pushState({}, 'Test page', route)

  return {
    user: userEvent.setup(),
    ...render((<Provider store={store}>{ui}</Provider>), {wrapper: BrowserRouter}),
  }
}
