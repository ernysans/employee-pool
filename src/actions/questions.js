import {setLoading} from "./loading";
import {_saveQuestion, _saveQuestionAnswer} from "../utils/_DATA";

export const RECEIVE_QUESTIONS = 'RECEIVE_QUESTIONS';
export const ADD_QUESTION = 'ADD_QUESTION';
export const ADD_ANSWER = 'ADD_ANSWER';

export function receiveQuestions(questions) {
  return {
    type: RECEIVE_QUESTIONS,
    questions,
  };
}

export function addQuestion(question) {
  return {
    type: ADD_QUESTION,
    question,
  };
}

export function addAnswer({authedUser, answer, qid}) {
  return {
    type: ADD_ANSWER,
    authedUser,
    qid,
    answer,
  };
}


/**
 * Handle vote
 * @param option
 * @param qid
 */
export const handleVote = ({option, qid}) => {
  return (dispatch, getState) => {
    const {authedUser, loading} = getState();
    if (loading) return;
    if (!option) return alert('Please select an option');
    dispatch(setLoading(true));
    // Save answer and update store
    return _saveQuestionAnswer({
      authedUser,
      qid: qid,
      answer: option,
    }).then(() => {
      dispatch(addAnswer({
        authedUser,
        qid: qid,
        answer: option,
      }));
      dispatch(setLoading(false));
    }).catch((e) => {
      dispatch(setLoading(false));
      alert('Something went wrong, please try again later:' + e.message);
    })
  }
};

export const handleAddQuestion = ({optionOneText, optionTwoText}) => {
  return (dispatch, getState) => {
    // const navigate = useNavigate();
    const {authedUser, loading} = getState();
    const allOk = optionOneText.length > 0 && optionTwoText.length > 0 && !loading;
    if (!allOk) {
      alert('Please fill all the fields');
      dispatch(setLoading(false));
      return;
    }
    if (loading) return;
    const question = {
      optionOneText: optionOneText,
      optionTwoText: optionTwoText,
      author: authedUser,
    }
    if (!optionOneText || !optionTwoText) return alert('Please fill in both options');
    dispatch(setLoading(true));
    // Save question and update store
    return _saveQuestion(question).then((formattedQuestion) => {
      dispatch(addQuestion(formattedQuestion));
      dispatch(setLoading(false));
      // navigate('/');
    }).catch((error) => {
      alert('Error creating question', error.message);
      dispatch(setLoading(false));
    });
  }
};
