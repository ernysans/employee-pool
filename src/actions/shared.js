import {getInitialData} from "../utils/api";
import {receiveUsers} from "./users";
import {receiveQuestions} from "./questions";
import {setLoading} from "./loading";

export function handleInitialData() {
  return (dispatch) => {
    dispatch(setLoading(true));
    return getInitialData().then(({users, questions}) => {
      dispatch(receiveUsers(users));
      dispatch(receiveQuestions(questions));
      dispatch(setLoading(false));
    });
  };
}
