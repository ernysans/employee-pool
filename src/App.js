import {Fragment, useEffect} from 'react';
import "material-components-web/dist/material-components-web.min.css";
import './styles/App.css';
import {Route, Routes} from "react-router-dom";
import DashboardPage from "./components/pages/DashboardPage";
import PoolPage from "./components/pages/PoolPage";
import PollCreationPage from "./components/pages/PollCreationPage";
import LeaderboardPage from "./components/pages/LeaderboardPage";
import {handleInitialData} from "./actions/shared";
import {connect} from "react-redux";
import LogInPage from "./components/pages/LogInPage";
import Progressbar from "./components/shared/Progressbar";

const App = ({dispatch, ready, authedUser}) => {
  useEffect(() => {
    handleInitialData()(dispatch);
  }, [dispatch]);
  if (!ready) return (<Progressbar/>);
  return (<div className="app">
    <Fragment>
      <Progressbar/>
      {!authedUser && (<LogInPage/>)}
      {authedUser && (<Routes>
        <Route exact path="/" element={<DashboardPage/>}/>
        <Route exact path="/add" element={<PollCreationPage/>}/>
        <Route path="/questions/:question_id" element={<PoolPage/>}/>
        <Route exact path="/leaderboard" element={<LeaderboardPage/>}/>
      </Routes>)}
    </Fragment>
  </div>);
};

const mapStateToProps = ({users, questions, authedUser}) => ({
  ready: Object.keys(users).length > 0 && Object.keys(questions).length > 0,
  authedUser,
});

export default connect(mapStateToProps)(App);
