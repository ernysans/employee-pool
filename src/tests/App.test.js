import React from 'react';
import {fireEvent, screen, waitFor} from '@testing-library/react';
import App from "../App";
import {renderWithRouter} from "../utils/test-utils";
import {store} from "../utils/store";
import {setAuthedUser} from "../actions/authedUser";

describe('App', () => {
  it('Login Page renders without crashing', async () => {
    const route = '/';
    renderWithRouter(<App/>, {route});
    await waitFor(() => {
      expect(screen.getByText(/Sign in/i)).toBeInTheDocument();
    }, 1000);
  });
  it('The navigation link selected on app load with custom route', async () => {
    store.dispatch(setAuthedUser('tester'));
    const route = '/add';
    renderWithRouter(<App/>, {route});
    const selected = screen.getByTestId('nav-button-' + route);
    expect(selected.getAttribute('aria-selected')).toBe('true');
  });
  it('link selected status changes on click', async () => {
    const route = '/';
    const targetRoute = '/leaderboard';
    renderWithRouter(<App/>, {route});
    const button = screen.getByTestId('nav-button-' + targetRoute);
    expect(button.getAttribute('aria-selected')).toBe('false');
    fireEvent.click(button);
    const selected = screen.getByTestId('nav-button-' + targetRoute);
    expect(selected.getAttribute('aria-selected')).toBe('true');
  });

  it('Sign out button works', async () => {
    const route = '/';
    renderWithRouter(<App/>, {route});
    const button = screen.getByTestId('sign-out-button');
    fireEvent.click(button);
    expect(screen.getByText(/Sign in/i)).toBeInTheDocument();
  });

  it('Snapshot test', async () => {
    const route = '/';
    const {container} = renderWithRouter(<App/>, {route});
    expect(container).toMatchSnapshot();
  });
});
