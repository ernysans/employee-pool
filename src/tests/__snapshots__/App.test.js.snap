// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Snapshot test 1`] = `
<div>
  <div
    class="app"
  >
    <div
      aria-label="Example Progress Bar"
      aria-valuemax="1"
      aria-valuemin="0"
      aria-valuenow="0"
      class="mdc-linear-progress mdc-linear-progress--indeterminate mdc-linear-progress--animation-ready"
      role="progressbar"
    >
      <div
        class="mdc-linear-progress__buffer"
      >
        <div
          class="mdc-linear-progress__buffer-bar"
        />
        <div
          class="mdc-linear-progress__buffer-dots"
        />
      </div>
      <div
        class="mdc-linear-progress__bar mdc-linear-progress__primary-bar"
      >
        <span
          class="mdc-linear-progress__bar-inner"
        />
      </div>
      <div
        class="mdc-linear-progress__bar mdc-linear-progress__secondary-bar"
      >
        <span
          class="mdc-linear-progress__bar-inner"
        />
      </div>
    </div>
    <div
      id="login-page"
    >
      <div
        class="container"
      >
        <div
          class="content-container"
        >
          <h1
            class="mdc-typography--headline4"
          >
            Employee Pools
          </h1>
          <h2
            class="mdc-typography--headline6"
          >
            Sign in
          </h2>
          <form>
            <div
              class="mdc-text-field mdc-text-field--filled"
            >
              <input
                autocomplete="off"
                class="mdc-text-field__input"
                data-testid="username"
                id="username"
                type="text"
                value=""
              />
              <div
                class="mdc-notched-outline"
              >
                <div
                  class="mdc-notched-outline__leading"
                />
                <div
                  class="mdc-notched-outline__notch"
                >
                  <label
                    class="mdc-floating-label mdc-floating-label--float-above	mdc-floating-label--required"
                    for="username"
                  >
                    Username
                  </label>
                </div>
                <div
                  class="mdc-notched-outline__trailing"
                />
              </div>
            </div>
            <div
              class="mdc-text-field mdc-text-field--filled"
            >
              <input
                autocomplete="off"
                class="mdc-text-field__input"
                data-testid="password"
                disabled=""
                id="password"
                type="password"
                value=""
              />
              <div
                class="mdc-notched-outline"
              >
                <div
                  class="mdc-notched-outline__leading"
                />
                <div
                  class="mdc-notched-outline__notch"
                >
                  <label
                    class="mdc-floating-label mdc-floating-label--float-above	mdc-floating-label--required mdc-floating-label--shake	"
                    for="password"
                  >
                    Password
                  </label>
                </div>
                <div
                  class="mdc-notched-outline__trailing"
                />
              </div>
            </div>
            <button
              class="mdc-button mdc-button--raised"
              data-testid="submit"
              disabled=""
            >
              <span
                class="mdc-button__label"
              >
                Login
              </span>
            </button>
          </form>
        </div>
      </div>
    </div>
  </div>
</div>
`;
